hail/src/main/scala/is/hail/expr/ir/Requiredness.scala

```diff
@@ -96,6 +96,14 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
 
   def addBindingRelations(node: BaseIR): Unit = {
     val refMap = usesAndDefs.uses(node).toArray.groupBy(_.t.name).mapValues(_.asInstanceOf[Array[RefEquality[BaseIR]]])
+    def addDeflessBinding(name: String, d: IR, req: BaseTypeWithRequiredness): Unit = {
+      if (refMap.contains(name)) {
+        val uses = refMap(name)
+        uses.foreach { u => defs.bind(u, Array(req)) }
+        dependents.getOrElseUpdate(d, mutable.Set[RefEquality[BaseIR]]()) ++= uses
```

Comment:
In this case, since the type of the `name` refs you're binding don't depend on any of the requiredness computation in the `ShuffleWith` node (They're always `TShuffle(required)`), they don't need to be added as dependents of the base (defining) node here since we don't need to recompute subtrees containing these refs if the ShuffleWith node changes. I think in this case I'd probably just inline this block in the ShuffleWith case, since it becomes a non-generic binding function.

Reply:
(It's not going to yield incorrect results if you leave it in, you may just spend time unnecessarily computing subtrees when nothing has changed in the underlying requirednesses)

Reply:
OK, addressed this.